From cc39da6a8bf9bbe67c7ca0da02cb7faf05294571 Mon Sep 17 00:00:00 2001
From: Stefan Berger <stefanb@linux.ibm.com>
Date: Wed, 26 Jan 2022 09:30:18 -0500
Subject: [PATCH 28/89] ima: Add template_name field to ima_config

Add a template_name field to the ima_config structure in preparation
for allowing users to choose the template name of an IMA namespace
vai a file in IMA's securityfs.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>
---
 security/integrity/ima/ima.h             | 2 ++
 security/integrity/ima/ima_init_ima_ns.c | 1 +
 security/integrity/ima/ima_template.c    | 4 ++++
 3 files changed, 7 insertions(+)

diff --git a/security/integrity/ima/ima.h b/security/integrity/ima/ima.h
index 587f0e9e5a49..11d34fc9b7e9 100644
--- a/security/integrity/ima/ima.h
+++ b/security/integrity/ima/ima.h
@@ -126,6 +126,8 @@ struct ima_h_table {
 struct ima_config {
 	int ima_hash_algo;
 	int hash_setup_done;
+
+	char template_name[32];
 };
 
 struct ima_namespace {
diff --git a/security/integrity/ima/ima_init_ima_ns.c b/security/integrity/ima/ima_init_ima_ns.c
index a7573f8ec6c8..1ac3f97a2d42 100644
--- a/security/integrity/ima/ima_init_ima_ns.c
+++ b/security/integrity/ima/ima_init_ima_ns.c
@@ -107,6 +107,7 @@ struct ima_namespace init_ima_ns = {
 	.ima_ns_flags = BIT(IMA_NS_ACTIVE),
 	.config = {
 		.ima_hash_algo = HASH_ALGO_SHA1,
+		.template_name = CONFIG_IMA_DEFAULT_TEMPLATE,
 	},
 };
 EXPORT_SYMBOL(init_ima_ns);
diff --git a/security/integrity/ima/ima_template.c b/security/integrity/ima/ima_template.c
index f46dbc307324..9b4fe9a732c2 100644
--- a/security/integrity/ima/ima_template.c
+++ b/security/integrity/ima/ima_template.c
@@ -135,6 +135,10 @@ static int __init ima_template_setup(char *str)
 
 	ns->ima_template = template_desc;
 	ns->template_setup_done = 1;
+
+	strscpy(ns->config.template_name, str,
+		sizeof(ns->config.template_name));
+
 	return 1;
 }
 __setup("ima_template=", ima_template_setup);
-- 
2.40.1

